packages/flutter/lib/src/widgets/text_editing_action.dart

```diff
@@ -0,0 +1,63 @@
+// Copyright 2014 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+import 'package:flutter/foundation.dart';
+import 'package:flutter/rendering.dart' show RenderEditable;
+
+import 'actions.dart';
+import 'editable_text.dart';
+import 'focus_manager.dart';
+import 'framework.dart';
+
+/// An implementor of this must be focused for a [TextEditingAction] to be
+/// enabled.
+///
+/// See also:
+///
+///   * [EditableText], which implements this and is the most typical target of
+///     a TextEditingAction.
+abstract class TextEditingActionTarget {
+  /// The renderer that handles [TextEditingAction]s.
+  ///
+  /// See also:
+  ///
+  /// * [EditableTextState.renderEditable], which overrides this.
+  RenderEditable get renderEditable;
+}
+
+/// An [Action] related to editing text.
+///
+/// Enables itself only when a [TextEditingActionTarget], e.g. [EditableText],
+/// is currently focused. The result of this is that when a
+/// TextEditingActionTarget is not focused, it will fall through to any
+/// non-TextEditingAction that handles the same shortcut. For example,
+/// overriding the tab key in [Shortcuts] with a TextEditingAction will only
+/// invoke your TextEditingAction when a TextEditingActionTarget is focused,
+/// otherwise the default tab behavior will apply.
+///
+/// The currently focused TextEditingActionTarget is available in the [invoke]
+/// method via [textEditingActionTarget].
+///
+/// See also:
+///
+///  * [CallbackAction], which is a similar Action type but unrelated to text
+///    editing.
+abstract class TextEditingAction<T extends Intent> extends ContextAction<T> {
+  /// Returns the currently focused [TextEditingAction], if any.
+  @protected
+  TextEditingActionTarget? get textEditingActionTarget {
+    // If a TextEditingActionTarget is not focused, then ignore this action.
+    if ((primaryFocus!.context! as StatefulElement).state is! TextEditingActionTarget) {
```

Comment:
is it safe to use `!` on `primaryFocus`, or assuming the context is a `StatefulElement`?

Reply:
Oh thanks!  I messed this up when I refactored it.  Pushing a fix.